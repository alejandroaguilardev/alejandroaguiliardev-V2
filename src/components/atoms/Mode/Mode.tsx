import { IconButton } from "@mui/material"
import { ToggleOn, ToggleOff, LightMode, DarkMode } from '@mui/icons-material';
import { useThemeContext } from "../../../context";
import { useEffect } from "react";


export const Mode = () => {
    const { theme, handleTheme } = useThemeContext();

    useEffect(() => {
        handleTheme(localStorage.getItem('mode') ?? 'light');
    }, [])

    return (
        <IconButton
            onClick={() => {
                handleTheme(theme.mode === 'dark' ? 'light' : 'dark');
                localStorage.setItem('mode', theme.mode === 'dark' ? 'light' : 'dark')
            }}
            sx={{ fontSize: "1rem" }}
        >
            {theme.mode === 'dark' ?
                <>
                    <DarkMode />
                    <ToggleOn sx={{ fontSize: "3rem", color: 'primary.main' }} />
                </>
                :
                <>
                    <LightMode />
                    <ToggleOff sx={{ fontSize: "3rem" }} />
                </>
            }
        </IconButton >
    )
}
