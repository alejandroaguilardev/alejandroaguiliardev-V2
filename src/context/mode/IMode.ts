import { ReactElement } from "react";

export interface IMode {
	mode: string;
}

export interface IModeProvider {
	children:  React.ReactElement | React.ReactElement[] | React.ReactNode;
}

export interface IInit {
	theme: { mode: string };
	handleTheme: (mode: string) => void;
}
