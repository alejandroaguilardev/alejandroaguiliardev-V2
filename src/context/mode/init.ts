export const INIT = {
    mode: 'light'
}