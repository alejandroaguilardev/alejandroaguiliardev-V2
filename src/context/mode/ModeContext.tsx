import { createContext } from "react";
import { IInit } from "./IMode";

export const ModeContext = createContext({} as IInit);
