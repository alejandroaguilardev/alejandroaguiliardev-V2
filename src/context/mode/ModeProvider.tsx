import { useEffect, useReducer } from "react";
import { IModeProvider } from "./IMode";
import { INIT } from "./init";
import { ModeContext } from "./ModeContext";
import { ModeReducer } from "./ModeReducer";

export const ModeProvider = ({ children }: IModeProvider) => {
	const [theme, dispatch] = useReducer(ModeReducer, INIT);

	const handleTheme = (mode: string) => {
		dispatch({
			type: "changeMode",
			payload: { mode },
		});
	};

	useEffect(() => {
		handleTheme(localStorage.getItem("mode") ?? "light");
	}, []);

	return (
		<ModeContext.Provider value={{ theme, handleTheme }}>
			{children}
		</ModeContext.Provider>
	);
};
