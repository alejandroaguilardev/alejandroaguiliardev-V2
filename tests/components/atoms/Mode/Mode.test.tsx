import { fireEvent, render, screen } from "@testing-library/react";
import { Mode } from "../../../../src/components/atoms/Mode/Mode";
import { INIT } from "../../../../src/context/mode/init";
import { ModeContext } from "../../../../src/context/mode/ModeContext";

describe("Mode test", () => {
	test("toMacthSnapshot", () => {
		const mockHandleTheme = jest.fn();
		const { container } = render(
			<ModeContext.Provider
				value={{ theme: INIT, handleTheme: mockHandleTheme }}
			>
				<Mode />
			</ModeContext.Provider>
		);
		const btn = screen.getByLabelText("icon-mode");
		fireEvent.click(btn);

		expect(container).toMatchSnapshot();
		expect(mockHandleTheme).toHaveBeenCalledWith("dark")
	});
});
