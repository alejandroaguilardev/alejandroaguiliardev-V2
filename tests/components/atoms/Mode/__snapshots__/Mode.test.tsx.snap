// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mode test toMacthSnapshot 1`] = `
<div>
  <button
    aria-label="icon-mode"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-q44og9-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="LightModeIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0 .39-.39.39-1.03 0-1.41l-1.06-1.06zm1.06-10.96c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tzp3k3-MuiSvgIcon-root"
      data-testid="ToggleOffIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M17 7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h10c2.76 0 5-2.24 5-5s-2.24-5-5-5zM7 15c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
